### PR TITLE
WIP: Target exporting with CMake

### PR DESCRIPTION
**Pull request by undisputed-seraphim**
This merge request aims to implement proper target exporting and importing with CMake.

I need to make these changes because certain embedded platform build systems expect `myproject-targets.cmake` files to be present in order to facilitate dependency resolution. Existing dependency layers that are bundled with the build system also expect other dependencies to be reasonably well-behaved (does not force-set CMake variables that leak across projects).

High-level overview of changes:
- md5 moved into its own project - This is because OpenRAVE uses only those symbols and it is unnecessary to link against the whole `cpp-gen-md5` binary just to get basic md5 symbols. This has to be changed because otherwise `cpp-gen-md5` also has to be exported and installed together with libopenrave.
- Added `EXPORT` to most `install()` commands.
- Use `find_package(Python2)` to set Python-related directories, and the now-deprecated `find_package(PythonLibs 2)` code is removed.
- Rewrote `openrave-config.cmake.in` to fit modern standards.
- `openrave-config-version.cmake` deleted - It is now automatically generated by CMake.
- `myrelpath.py` deleted - It is used solely to determine the prefix path of the installation directory of python modules, but CMake has a built-in function for this now.
- Adjusted the linking visibility of some libraries so they can be comfortably exported (CMake throws an error otherwise)
- Use version agnostic FindPython, and converted the linking syntax from old CMake 2.x syntax to the CMake 3.x target-based syntax. (Note that support for <= CMake 2.8.12 will be dropped in a future release.)

There is also a compatibility shim so that existing libraries that use deprecated behaviours continue to work, but ideally we should fix downstream libraries and them remove this shim.

After this change, any project that wishes to link against OpenRAVE in CMake do not need any to do any voodoo; simply `find_package(OpenRAVE REQUIRED)` and then `target_link_libraries(myproject PUBLIC OpenRAVE::OpenRAVE)`.

### CODE REVIEW

**Comment by cielavenir** [comment]
oh please check if this change is still compatible with Python3 flavor

**Comment by undisputed-seraphim** [comment]
> oh please check if this change is still compatible with Python3 flavor

The change is purely CMake only, to target Python3 we just change all `find_package(Python2 ...)` to `find_package(Python3 ...)`

File: openrave-config.cmake.in [hunk 1]
```diff
@@ -32,76 +11,40 @@
...
-set( OpenRAVE_Boost_VERSION "@Boost_MAJOR_VERSION@.@Boost_MINOR_VERSION@")
-set( OpenRAVE_BLA_VENDOR "@BLA_VENDOR@" )
+find_package(Boost ${OpenRAVE_Boost_VERSION} REQUIRED COMPONENTS date_time filesystem system thread)
+find_package(RapidJSON REQUIRED)
+if (${OpenRAVE_PYTHON})
+    find_package(Python2 REQUIRED COMPONENTS Interpreter Development NumPy)
+
+    # Backward compatibility shim. Remove this line once we finally migrate to modern CMake.
+    # Ideally we should be linking against the target Python2::Development to gain access to headers, instead of doing this.
+    set(PYTHON_INCLUDE_DIR ${Python2_INCLUDE_DIRS} CACHE STRING "This is for backward compatibility with deprecated CMake declarations." FORCE)
```
Comment (cielavenir) [inline openrave-config.cmake.in:35]: should be OpenRAVE_PYTHON_INCLUDE_DIRS, not PYTHON_INCLUDE_DIR

**Comment by cielavenir** [comment]
> to target Python3 we just change all find_package(Python2 ...) to find_package(Python3 ...)

we will need to patch a lot.

Also, Currently the behavior can be changed by installing `python-is-python2` or `python-is-python3` without changing CMakeLists, but we will lose that capability.

**Comment by undisputed-seraphim** [comment]
> > to target Python3 we just change all find_package(Python2 ...) to find_package(Python3 ...)
> 
> we will need to patch a lot.
> 
> Also, Currently the behavior can be changed by installing `python-is-python2` or `python-is-python3` without changing CMakeLists, but we will lose that capability.

We can use [FindPython](https://cmake.org/cmake/help/latest/module/FindPython.html) , which will define target names that are version agnostic. i.e.,

```
find_package(Python 2.7 REQUIRED COMPONENTS Interpreter Development NumPy)
...
target_link_libraries(my_library PUBLIC Python::Module Python::NumPy)
```

That way, we only need to change `2.7` to `3.x` without having to change anything else.

Does that work for you?

**Comment by undisputed-seraphim** [comment]
Closing this due to developments on another branch involving CMake.

**Pull request closed without merging**